chain/market/fundmanager.go

```diff
@@ -0,0 +1,600 @@
+	StateWaitMsg(ctx context.Context, cid cid.Cid, confidence uint64) (*api.MsgLookup, error)
+}
+
+// FundManager keeps track of funds in a set of addresses
+type FundManager struct {
+	ctx      context.Context
+	shutdown context.CancelFunc
+	api      fundManagerAPI
+	wallet   address.Address
+	str      *Store
+
+	lk          sync.Mutex
+	fundedAddrs map[address.Address]*fundedAddress
+}
+
+type waitSentinel cid.Cid
+
+var waitSentinelUndef = waitSentinel(cid.Undef)
+
+func NewFundManager(api fundManagerAPI, ds datastore.Batching, wallet address.Address) *FundManager {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &FundManager{
+		ctx:         ctx,
+		shutdown:    cancel,
+		api:         api,
+		wallet:      wallet,
+		str:         newStore(ds),
+		fundedAddrs: make(map[address.Address]*fundedAddress),
+	}
+}
+
+func (fm *FundManager) Stop() {
+	fm.shutdown()
+}
+
+func (fm *FundManager) Start() error {
+	fm.lk.Lock()
+	defer fm.lk.Unlock()
+
+	// TODO:
+	// To save memory:
+	// - in State() only load addresses with in-progress messages
+	// - load the others just-in-time from getFundedAddress
+	// - delete(fm.fundedAddrs, addr) when the queue has been processed
+	return fm.str.forEach(func(state *FundedAddressState) {
+		fa := newFundedAddress(fm, state.Addr)
+		fa.state = state
+		fm.fundedAddrs[fa.state.Addr] = fa
+		fa.start()
+	})
+}
+
+// Creates a fundedAddress if it doesn't already exist, and returns it
+func (fm *FundManager) getFundedAddress(addr address.Address) *fundedAddress {
+	fm.lk.Lock()
+	defer fm.lk.Unlock()
+
+	fa, ok := fm.fundedAddrs[addr]
+	if !ok {
+		fa = newFundedAddress(fm, addr)
+		fm.fundedAddrs[addr] = fa
+	}
+	return fa
+}
+
+// Reserve adds amt to `reserved`. If there is not enough available funds for
+// the address, submits a message on chain to top up available funds.
+func (fm *FundManager) Reserve(ctx context.Context, addr address.Address, amt abi.TokenAmount) (waitSentinel, error) {
+	return fm.getFundedAddress(addr).reserve(ctx, amt)
+}
+
+// Subtract from `reserved`.
+func (fm *FundManager) Release(ctx context.Context, addr address.Address, amt abi.TokenAmount) error {
+	return fm.getFundedAddress(addr).release(ctx, amt)
+}
+
+// Withdraw unreserved funds. Only succeeds if there are enough unreserved
+// funds for the address.
+func (fm *FundManager) Withdraw(ctx context.Context, addr address.Address, amt abi.TokenAmount) (waitSentinel, error) {
+	return fm.getFundedAddress(addr).withdraw(ctx, amt)
+}
+
+// Waits for a reserve or withdraw to complete.
+func (fm *FundManager) Wait(ctx context.Context, sentinel waitSentinel) error {
+	_, err := fm.api.StateWaitMsg(ctx, cid.Cid(sentinel), build.MessageConfidence)
+	return err
+}
+
+// FundedAddressState keeps track of the state of an address with funds in the
+// datastore
+type FundedAddressState struct {
+	// Wallet is the wallet from which funds are added to the address
+	Wallet address.Address
```

Comment:
I actually don't think we need a  wallet here. What is reserved is funds that are in Escrow for Addr, but not locked. What wallet transferred funds in really doesn't matter

Reply:
(what wallet withdraws DOES matter in that a withdraw message won't go through if the wallet isn't either the same as the address or the owner/worker address in the case of a miner, but could potentially be a different one than transferred funds in)